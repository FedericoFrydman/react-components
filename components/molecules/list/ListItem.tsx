import { forwardRef, Ref } from 'react';
import cn from 'classnames';

import { Text, TextProps } from '../../../components/atoms/text/Text';
import './list-item.scss';
import { Button, ButtonProps } from '../button/Button';

const classListItem = 'lsux-list-item';

export interface ListItemTextProps {
  primary?: string;
  primaryProps?: TextProps;
  secondary?: string;
  secondaryProps?: TextProps;
  textAlign?: 'left' | 'center' | 'right';
}

export const ListItemText = ({
  primary,
  primaryProps,
  secondary,
  secondaryProps,
  textAlign = 'left',
}: ListItemTextProps) => {
  return (
    <div className={cn(`${classListItem}__text`, `${classListItem}__text--${textAlign}`)}>
      <div>
        <div>
          <Text
            {...primaryProps}
            textSize={primaryProps?.textSize ?? 'large'}
            classNames={[`${classListItem}__text-primary`, primaryProps?.classNames]}
          >
            {primary}
          </Text>
        </div>
        <div>
          <Text
            {...secondaryProps}
            textSize={secondaryProps?.textSize ?? 'small'}
            classNames={[`${classListItem}__text-secondary`, secondaryProps?.classNames]}
          >
            {secondary}
          </Text>
        </div>
      </div>
    </div>
  );
};

export interface ListItemButtonProps extends ButtonProps {
  children?: string;
}

export const ListItemButton = ({ children, ...props }: ListItemButtonProps) => {
  return (
    <Button
      {...props}
      label={children ?? props.label}
      classNames={[`${classListItem}__button`]}
      variant="secondary"
      stretch
    />
  );
};

interface CompoundedListItemComponent
  extends React.ForwardRefExoticComponent<ListItemProps & React.RefAttributes<HTMLLIElement>> {
  Button?: React.FC<ListItemButtonProps>;
  Text?: React.FC<ListItemTextProps>;
}

export interface ListItemProps {
  /**
   * The children of the ListItem.
   */
  children?: React.ReactNode;
  /**
   * Additional classNames to add to the list item.
   * @default []
   */
  classNames?: string[];
  /**
   * The ref to the HTML DOM element.
   */
  ref?: Ref<HTMLLIElement>;
}

export const ListItem = forwardRef<HTMLLIElement, ListItemProps>(
  ({ children, classNames = [], ...props }: ListItemProps, ref: ListItemProps['ref']) => {
    const cnListItem = cn('lsux-list-item', classNames);
    return (
      <li {...props} ref={ref} className={cnListItem}>
        {children}
      </li>
    );
  },
) as CompoundedListItemComponent;

// eslint-disable-next-line react/display-name
ListItem.Button = (props) => <ListItemButton {...props} />;
// eslint-disable-next-line react/display-name
ListItem.Text = (props) => <ListItemText {...props} />;

ListItem.displayName = 'ListItem';
