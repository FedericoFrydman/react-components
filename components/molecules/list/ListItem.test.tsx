import { shallow, ShallowWrapper } from 'enzyme';
import { render, screen } from '@testing-library/react';
import cn from 'classnames';

import { ListItem, ListItemProps } from './ListItem';

describe('ListItem Component', () => {
  let testId: string;
  let wrapper: ShallowWrapper;

  beforeEach(() => {
    testId = 'test list item';
    wrapper = shallow(<ListItem>{testId}</ListItem>);
  });

  describe('props', () => {
    describe('classNames', () => {
      describe('when provided classNames', () => {
        const classNames = ['my-class-1', 'my-class-2'];

        it('gives the list the provided classNames', () => {
          const listItem = renderListItem({ classNames });

          expect(listItem).toHaveClass(cn(classNames), { exact: false });
        });
      });
    });

    // describe('noBorder', () => {
    //   it('should contain border class when false', () => {
    //     const listItem = renderListItem();
    //     expect(listItem).toHaveClass('lsux-list--border');
    //   });
    //   it('should not contain border class when true', () => {
    //     const list = renderListItem({ noBorder: true });
    //     expect(list).not.toHaveClass('lsux-list--border');
    //   });
    // });
  });

  test('should render the component', () => {
    expect(wrapper).toMatchSnapshot();
  });

  const renderListItem = (props: ListItemProps = {}): HTMLElement => {
    render(<ListItem {...props} data-testid={testId} />);

    return screen.queryByTestId(testId);
  };
});
