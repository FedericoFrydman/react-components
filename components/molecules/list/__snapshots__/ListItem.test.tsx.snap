// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ListItem Component should render the component 1`] = `
<li
  className="lsux-list-item"
>
  test list item
</li>
`;
